File: java/admob/NativeAdvancedExample/app/src/main/java/com/google/example/gms/nativeadvancedexample/MainActivity.java
/*
 * Copyright (C) 2026 Google, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.example.gms.nativeadvancedexample;

import android.annotation.SuppressLint;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdLoader;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.LoadAdError;
import com.google.android.gms.ads.MobileAds;
import com.google.android.gms.ads.RequestConfiguration;
import com.google.android.gms.ads.VideoController;
import com.google.android.gms.ads.VideoOptions;
import com.google.android.gms.ads.nativead.NativeAd;
import com.google.android.gms.ads.nativead.NativeAdOptions;
import com.google.android.gms.ads.nativead.NativeAdView;
import com.google.example.gms.nativeadvancedexample.databinding.ActivityMainBinding;
import com.google.example.gms.nativeadvancedexample.databinding.AdUnifiedBinding;
import java.util.Arrays;
import java.util.Locale;
import java.util.concurrent.atomic.AtomicBoolean;

/** A simple activity class that displays native ad formats. */
@SuppressLint("SetTextI18n")
public class MainActivity extends AppCompatActivity {

  // Check your logcat output for the test device hashed ID e.g.
  // "Use RequestConfiguration.Builder().setTestDeviceIds(Arrays.asList("ABCDEF012345"))
  // to get test ads on this device" or
  // "Use new ConsentDebugSettings.Builder().addTestDeviceHashedId("ABCDEF012345") to set this as
  // a debug device".
  public static final String TEST_DEVICE_HASHED_ID = "ABCDEF012345";

  private static final String ADMOB_AD_UNIT_ID = "ca-app-pub-3940256099942544/2247696110";
  private static final String TAG = "MainActivity";

  private final AtomicBoolean isMobileAdsInitializeCalled = new AtomicBoolean(false);
  private GoogleMobileAdsConsentManager googleMobileAdsConsentManager;
  private ActivityMainBinding binding;
  private NativeAd nativeAd;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    binding = ActivityMainBinding.inflate(getLayoutInflater());
    setContentView(binding.getRoot());
    setSupportActionBar(binding.toolBar);

    // Log the Mobile Ads SDK version.
    Log.d(TAG, "Google Mobile Ads SDK Version: " + MobileAds.getVersion());

    googleMobileAdsConsentManager =
        GoogleMobileAdsConsentManager.getInstance(getApplicationContext());
    googleMobileAdsConsentManager.gatherConsent(
        this,
        consentError -> {
          if (consentError != null) {
            // Consent not obtained in current session.
            Log.w(
                TAG,
                String.format("%s: %s", consentError.getErrorCode(), consentError.getMessage()));
          }

          if (googleMobileAdsConsentManager.canRequestAds()) {
            binding.refreshButton.setVisibility(View.VISIBLE);
            initializeMobileAdsSdk();
          }

          if (googleMobileAdsConsentManager.isPrivacyOptionsRequired()) {
            // Regenerate the options menu to include a privacy setting.
            invalidateOptionsMenu();
          }
        });

    // This sample attempts to load ads using consent obtained in the previous session.
    if (googleMobileAdsConsentManager.canRequestAds()) {
      initializeMobileAdsSdk();
    }
    binding.refreshButton.setOnClickListener(
        new View.OnClickListener() {
          @Override
          public void onClick(View unusedView) {
            if (googleMobileAdsConsentManager.canRequestAds()) {
              refreshAd();
            }
          }
        });
  }

  @Override
  public boolean onCreateOptionsMenu(Menu menu) {
    getMenuInflater().inflate(R.menu.action_menu, menu);
    MenuItem privacyItem = menu.findItem(R.id.privacy_settings);
    if (privacyItem != null) {
      privacyItem.setVisible(googleMobileAdsConsentManager.isPrivacyOptionsRequired());
    }
    return true;
  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    if (item.getItemId() == R.id.privacy_settings) {
      // Handle changes to user consent.
      googleMobileAdsConsentManager.showPrivacyOptionsForm(
          this,
          formError -> {
            if (formError != null) {
              Toast.makeText(this, formError.getMessage(), Toast.LENGTH_SHORT).show();
            }
          });
      return true;
    } else if (item.getItemId() == R.id.ad_inspector) {
      MobileAds.openAdInspector(
          this,
          error -> {
            // Error will be non-null if ad inspector closed due to an error.
            if (error != null) {
              Toast.makeText(this, error.getMessage(), Toast.LENGTH_SHORT).show();
            }
          });
      return true;
    }
    return super.onOptionsItemSelected(item);
  }

  /**
   * Populates a {@link NativeAdView} object with data from a given {@link NativeAd}.
   *
   * @param nativeAd the object containing the ad's assets
   * @param adBinding the binding object of the ad view
   */
  private void populateNativeAdView(NativeAd nativeAd, AdUnifiedBinding adBinding) {
    NativeAdView adView = (NativeAdView) adBinding.getRoot();

    // Set the media view.
    adView.setMediaView(adBinding.adMedia);

    // Set other ad assets.
    adView.setHeadlineView(adBinding.adHeadline);
    adView.setBodyView(adBinding.adBody);
    adView.setCallToActionView(adBinding.adCallToAction);
    adView.setIconView(adBinding.adAppIcon);
    adView.setPriceView(adBinding.adPrice);
    adView.setStarRatingView(adBinding.adStars);
    adView.setStoreView(adBinding.adStore);
    adView.setAdvertiserView(adBinding.adAdvertiser);

    // The headline and mediaContent are guaranteed to be in every NativeAd.
    adBinding.adHeadline.setText(nativeAd.getHeadline());
    adBinding.adMedia.setMediaContent(nativeAd.getMediaContent());

    // These assets aren't guaranteed to be in every NativeAd, so it's important to
    // check before trying to display them.
    if (nativeAd.getBody() == null) {
      adBinding.adBody.setVisibility(View.INVISIBLE);
    } else {
      adBinding.adBody.setVisibility(View.VISIBLE);
      adBinding.adBody.setText(nativeAd.getBody());
    }

    if (nativeAd.getCallToAction() == null) {
      adBinding.adCallToAction.setVisibility(View.INVISIBLE);
    } else {
      adBinding.adCallToAction.setVisibility(View.VISIBLE);
      adBinding.adCallToAction.setText(nativeAd.getCallToAction());
    }

    if (nativeAd.getIcon() == null) {
      adBinding.adAppIcon.setVisibility(View.GONE);
    } else {
      adBinding.adAppIcon.setImageDrawable(nativeAd.getIcon().getDrawable());
      adBinding.adAppIcon.setVisibility(View.VISIBLE);
    }

    if (nativeAd.getPrice() == null) {
      adBinding.adPrice.setVisibility(View.INVISIBLE);
    } else {
      adBinding.adPrice.setVisibility(View.VISIBLE);
      adBinding.adPrice.setText(nativeAd.getPrice());
    }

    if (nativeAd.getStore() == null) {
      adBinding.adStore.setVisibility(View.INVISIBLE);
    } else {
      adBinding.adStore.setVisibility(View.VISIBLE);
      adBinding.adStore.setText(nativeAd.getStore());
    }

    if (nativeAd.getStarRating() == null) {
      adBinding.adStars.setVisibility(View.INVISIBLE);
    } else {
      adBinding.adStars.setRating(nativeAd.getStarRating().floatValue());
      adBinding.adStars.setVisibility(View.VISIBLE);
    }

    if (nativeAd.getAdvertiser() == null) {
      adBinding.adAdvertiser.setVisibility(View.INVISIBLE);
    } else {
      adBinding.adAdvertiser.setText(nativeAd.getAdvertiser());
      adBinding.adAdvertiser.setVisibility(View.VISIBLE);
    }

    // This method tells the Google Mobile Ads SDK that you have finished populating your
    // native ad view with this native ad.
    adView.setNativeAd(nativeAd);

    // Get the video controller for the ad. One will always be provided, even if the ad doesn't
    // have a video asset.
    VideoController vc = nativeAd.getMediaContent().getVideoController();

    // Updates the UI to say whether or not this ad has a video asset.
    if (nativeAd.getMediaContent() != null && nativeAd.getMediaContent().hasVideoContent()) {

      // Create a new VideoLifecycleCallbacks object and pass it to the VideoController. The
      // VideoController will call methods on this object when events occur in the video
      // lifecycle.
      vc.setVideoLifecycleCallbacks(
          new VideoController.VideoLifecycleCallbacks() {
            @Override
            public void onVideoEnd() {
              // Publishers should allow native ads to complete video playback before
              // refreshing or replacing them with another ad in the same UI location.
              binding.refreshButton.setEnabled(true);
              binding.videoStatusText.setText("Video status: Video playback has ended.");
              super.onVideoEnd();
            }
          });
    } else {
      binding.videoStatusText.setText("Video status: Ad does not contain a video asset.");
      binding.refreshButton.setEnabled(true);
    }
  }

  /**
   * Creates a request for a new native ad based on the boolean parameters and calls the
   * corresponding "populate" method when one is successfully returned.
   */
  private void refreshAd() {
    binding.refreshButton.setEnabled(false);
    binding.videoStatusText.setText("");

    boolean startMuted = binding.startMutedCheckbox.isChecked();

    // It is recommended to call AdLoader.Builder on a background thread.
    new Thread(
            () -> {
              AdLoader.Builder builder = new AdLoader.Builder(this, ADMOB_AD_UNIT_ID);

              builder.forNativeAd(
                  new NativeAd.OnNativeAdLoadedListener() {
                    // OnLoadedListener implementation.
                    @Override
                    public void onNativeAdLoaded(NativeAd nativeAd) {
                      // If this callback occurs after the activity is destroyed, you must call
                      // destroy and return or you may get a memory leak.
                      boolean isDestroyed = false;
                      binding.refreshButton.setEnabled(true);
                      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
                        isDestroyed = isDestroyed();
                      }
                      if (isDestroyed || isFinishing() || isChangingConfigurations()) {
                        nativeAd.destroy();
                        return;
                      }
                      // You must call destroy on old ads when you are done with them,
                      // otherwise you will have a memory leak.
                      if (MainActivity.this.nativeAd != null) {
                        MainActivity.this.nativeAd.destroy();
                      }
                      MainActivity.this.nativeAd = nativeAd;
                      AdUnifiedBinding adBinding =
                          AdUnifiedBinding.inflate(getLayoutInflater(), binding.adFrame, false);
                      populateNativeAdView(nativeAd, adBinding);
                      binding.adFrame.removeAllViews();
                      binding.adFrame.addView(adBinding.getRoot());
                    }
                  });

              VideoOptions videoOptions =
                  new VideoOptions.Builder().setStartMuted(startMuted).build();

              NativeAdOptions adOptions =
                  new NativeAdOptions.Builder().setVideoOptions(videoOptions).build();

              builder.withNativeAdOptions(adOptions);

              AdLoader adLoader =
                  builder
                      .withAdListener(
                          new AdListener() {
                            @Override
                            public void onAdFailedToLoad(LoadAdError loadAdError) {
                              binding.refreshButton.setEnabled(true);
                              String error =
                                  String.format(
                                      Locale.getDefault(),
                                      "domain: %s, code: %d, message: %s",
                                      loadAdError.getDomain(),
                                      loadAdError.getCode(),
                                      loadAdError.getMessage());
                              Toast.makeText(
                                      MainActivity.this,
                                      "Failed to load native ad with error " + error,
                                      Toast.LENGTH_SHORT)
                                  .show();
                            }
                          })
                      .build();

              adLoader.loadAd(new AdRequest.Builder().build());
            })
        .start();
  }

  private void initializeMobileAdsSdk() {
    if (isMobileAdsInitializeCalled.getAndSet(true)) {
      return;
    }

    // Set your test devices.
    MobileAds.setRequestConfiguration(
        new RequestConfiguration.Builder()
            .setTestDeviceIds(Arrays.asList(TEST_DEVICE_HASHED_ID))
            .build());

    new Thread(
            () -> {
              // Initialize the Google Mobile Ads SDK on a background thread.
              MobileAds.initialize(this, initializationStatus -> {});

              // Load an ad on the main thread.
              runOnUiThread(() -> refreshAd());
            })
        .start();
  }

  @Override
  protected void onDestroy() {
    if (nativeAd != null) {
      nativeAd.destroy();
    }
    super.onDestroy();
  }
}

File: kotlin/admob/NativeAdvancedExample/app/src/main/java/com/google/android/gms/example/nativeadvancedexample/MainActivity.kt
/*
 * Copyright (C) 2017 Google, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.android.gms.example.nativeadvancedexample

import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.PopupMenu
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdLoader
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.MobileAds
import com.google.android.gms.ads.VideoController
import com.google.android.gms.ads.VideoOptions
import com.google.android.gms.ads.nativead.NativeAd
import com.google.android.gms.ads.nativead.NativeAdOptions
import com.google.android.gms.ads.nativead.NativeAdView
import com.google.android.gms.example.nativeadvancedexample.databinding.ActivityMainBinding
import com.google.android.gms.example.nativeadvancedexample.databinding.AdUnifiedBinding
import java.util.concurrent.atomic.AtomicBoolean

private const val TAG = "MainActivity"
const val ADMOB_AD_UNIT_ID = "ca-app-pub-3940256099942544/2247696110"

/** A simple activity class that displays native ad formats. */
class MainActivity : AppCompatActivity() {

  private val isMobileAdsInitializeCalled = AtomicBoolean(false)
  private lateinit var mainActivityBinding: ActivityMainBinding
  private lateinit var googleMobileAdsConsentManager: GoogleMobileAdsConsentManager
  private var currentNativeAd: NativeAd? = null

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    mainActivityBinding = ActivityMainBinding.inflate(layoutInflater)
    setContentView(mainActivityBinding.root)

    // Log the Mobile Ads SDK version.
    Log.d(TAG, "Google Mobile Ads SDK Version: " + MobileAds.getVersion())

    googleMobileAdsConsentManager = GoogleMobileAdsConsentManager(this)
    googleMobileAdsConsentManager.gatherConsent { consentError ->
      if (consentError != null) {
        // Consent not obtained in current session.
        Log.w(TAG, "${consentError.errorCode}. ${consentError.message}")
      }

      if (googleMobileAdsConsentManager.canRequestAds) {
        mainActivityBinding.refreshButton.visibility = View.VISIBLE
        initializeMobileAdsSdk()
      }

      if (googleMobileAdsConsentManager.isPrivacyOptionsRequired) {
        // Regenerate the options menu to include a privacy setting.
        invalidateOptionsMenu()
      }
    }

    // This sample attempts to load ads using consent obtained in the previous session.
    if (googleMobileAdsConsentManager.canRequestAds) {
      initializeMobileAdsSdk()
    }

    mainActivityBinding.refreshButton.setOnClickListener {
      if (googleMobileAdsConsentManager.canRequestAds) {
        refreshAd()
      }
    }
  }

  override fun onCreateOptionsMenu(menu: Menu): Boolean {
    menuInflater.inflate(R.menu.action_menu, menu)
    menu.findItem(R.id.action_more)?.apply {
      isVisible = googleMobileAdsConsentManager.isPrivacyOptionsRequired
    }
    return super.onCreateOptionsMenu(menu)
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean {
    val menuItemView = findViewById<View>(item.itemId)
    val activity = this
    PopupMenu(this, menuItemView).apply {
      menuInflater.inflate(R.menu.popup_menu, menu)
      show()
      setOnMenuItemClickListener { popupMenuItem ->
        when (popupMenuItem.itemId) {
          R.id.privacy_settings -> {
            // Handle changes to user consent.
            googleMobileAdsConsentManager.showPrivacyOptionsForm(activity) { formError ->
              if (formError != null) {
                Toast.makeText(this@MainActivity, formError.message, Toast.LENGTH_SHORT).show()
              }
            }
            true
          }
          else -> false
        }
      }
    }
    return super.onOptionsItemSelected(item)
  }

  /**
   * Populates a [NativeAdView] object with data from a given [NativeAd].
   *
   * @param nativeAd the object containing the ad's assets
   * @param unifiedAdBinding the binding object of the layout that has NativeAdView as the root view
   */
  private fun populateNativeAdView(nativeAd: NativeAd, unifiedAdBinding: AdUnifiedBinding) {
    val nativeAdView = unifiedAdBinding.root

    // Set the media view.
    nativeAdView.mediaView = unifiedAdBinding.adMedia

    // Set other ad assets.
    nativeAdView.headlineView = unifiedAdBinding.adHeadline
    nativeAdView.bodyView = unifiedAdBinding.adBody
    nativeAdView.callToActionView = unifiedAdBinding.adCallToAction
    nativeAdView.iconView = unifiedAdBinding.adAppIcon
    nativeAdView.priceView = unifiedAdBinding.adPrice
    nativeAdView.starRatingView = unifiedAdBinding.adStars
    nativeAdView.storeView = unifiedAdBinding.adStore
    nativeAdView.advertiserView = unifiedAdBinding.adAdvertiser

    // The headline and media content are guaranteed to be in every UnifiedNativeAd.
    unifiedAdBinding.adHeadline.text = nativeAd.headline
    nativeAd.mediaContent?.let { unifiedAdBinding.adMedia.setMediaContent(it) }

    // These assets aren't guaranteed to be in every UnifiedNativeAd, so it's important to
    // check before trying to display them.
    if (nativeAd.body == null) {
      unifiedAdBinding.adBody.visibility = View.INVISIBLE
    } else {
      unifiedAdBinding.adBody.visibility = View.VISIBLE
      unifiedAdBinding.adBody.text = nativeAd.body
    }

    if (nativeAd.callToAction == null) {
      unifiedAdBinding.adCallToAction.visibility = View.INVISIBLE
    } else {
      unifiedAdBinding.adCallToAction.visibility = View.VISIBLE
      unifiedAdBinding.adCallToAction.text = nativeAd.callToAction
    }

    if (nativeAd.icon == null) {
      unifiedAdBinding.adAppIcon.visibility = View.GONE
    } else {
      unifiedAdBinding.adAppIcon.setImageDrawable(nativeAd.icon?.drawable)
      unifiedAdBinding.adAppIcon.visibility = View.VISIBLE
    }

    if (nativeAd.price == null) {
      unifiedAdBinding.adPrice.visibility = View.INVISIBLE
    } else {
      unifiedAdBinding.adPrice.visibility = View.VISIBLE
      unifiedAdBinding.adPrice.text = nativeAd.price
    }

    if (nativeAd.store == null) {
      unifiedAdBinding.adStore.visibility = View.INVISIBLE
    } else {
      unifiedAdBinding.adStore.visibility = View.VISIBLE
      unifiedAdBinding.adStore.text = nativeAd.store
    }

    if (nativeAd.starRating == null) {
      unifiedAdBinding.adStars.visibility = View.INVISIBLE
    } else {
      unifiedAdBinding.adStars.rating = nativeAd.starRating!!.toFloat()
      unifiedAdBinding.adStars.visibility = View.VISIBLE
    }

    if (nativeAd.advertiser == null) {
      unifiedAdBinding.adAdvertiser.visibility = View.INVISIBLE
    } else {
      unifiedAdBinding.adAdvertiser.text = nativeAd.advertiser
      unifiedAdBinding.adAdvertiser.visibility = View.VISIBLE
    }

    // This method tells the Google Mobile Ads SDK that you have finished populating your
    // native ad view with this native ad.
    nativeAdView.setNativeAd(nativeAd)

    // Get the video controller for the ad. One will always be provided, even if the ad doesn't
    // have a video asset.
    val mediaContent = nativeAd.mediaContent
    val vc = mediaContent?.videoController

    // Updates the UI to say whether or not this ad has a video asset.
    if (vc != null && mediaContent.hasVideoContent()) {
      // Create a new VideoLifecycleCallbacks object and pass it to the VideoController. The
      // VideoController will call methods on this object when events occur in the video
      // lifecycle.
      vc.videoLifecycleCallbacks =
        object : VideoController.VideoLifecycleCallbacks() {
          override fun onVideoEnd() {
            // Publishers should allow native ads to complete video playback before
            // refreshing or replacing them with another ad in the same UI location.
            mainActivityBinding.refreshButton.isEnabled = true
            mainActivityBinding.videostatusText.text = "Video status: Video playback has ended."
            super.onVideoEnd()
          }
        }
    } else {
      mainActivityBinding.videostatusText.text = "Video status: Ad does not contain a video asset."
      mainActivityBinding.refreshButton.isEnabled = true
    }
  }

  /**
   * Creates a request for a new native ad based on the boolean parameters and calls the
   * corresponding "populate" method when one is successfully returned.
   */
  private fun refreshAd() {
    mainActivityBinding.refreshButton.isEnabled = false

    val builder = AdLoader.Builder(this, ADMOB_AD_UNIT_ID)

    builder.forNativeAd { nativeAd ->
      // OnUnifiedNativeAdLoadedListener implementation.
      // If this callback occurs after the activity is destroyed, you must call
      // destroy and return or you may get a memory leak.
      var activityDestroyed = false
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
        activityDestroyed = isDestroyed
      }
      if (activityDestroyed || isFinishing || isChangingConfigurations) {
        nativeAd.destroy()
        return@forNativeAd
      }
      // You must call destroy on old ads when you are done with them,
      // otherwise you will have a memory leak.
      currentNativeAd?.destroy()
      currentNativeAd = nativeAd
      val unifiedAdBinding = AdUnifiedBinding.inflate(layoutInflater)
      populateNativeAdView(nativeAd, unifiedAdBinding)
      mainActivityBinding.adFrame.removeAllViews()
      mainActivityBinding.adFrame.addView(unifiedAdBinding.root)
    }

    val videoOptions =
      VideoOptions.Builder().setStartMuted(mainActivityBinding.startMutedCheckbox.isChecked).build()

    val adOptions = NativeAdOptions.Builder().setVideoOptions(videoOptions).build()

    builder.withNativeAdOptions(adOptions)

    val adLoader =
      builder
        .withAdListener(
          object : AdListener() {
            override fun onAdFailedToLoad(loadAdError: LoadAdError) {
              val error =
                """
           domain: ${loadAdError.domain}, code: ${loadAdError.code}, message: ${loadAdError.message}
          """"
              mainActivityBinding.refreshButton.isEnabled = true
              Toast.makeText(
                  this@MainActivity,
                  "Failed to load native ad with error $error",
                  Toast.LENGTH_SHORT
                )
                .show()
            }
          }
        )
        .build()

    adLoader.loadAd(AdRequest.Builder().build())

    mainActivityBinding.videostatusText.text = ""
  }

  private fun initializeMobileAdsSdk() {
    if (isMobileAdsInitializeCalled.getAndSet(true)) {
      return
    }

    // Initialize the Mobile Ads SDK.
    MobileAds.initialize(this) { initializationStatus ->
      // Load an ad.
      refreshAd()
    }
  }

  override fun onDestroy() {
    currentNativeAd?.destroy()
    super.onDestroy()
  }
}
